
# Dokumentation


## Metadata

- R, Rstudio etc. , Version 
- Projektinformationen



## Reproduction of original results [1], Vorgehensweise, Ordnerstruktur

- 4 Sections unterteilt, Data, exec, inst, R
- 



## required Packages

- auflisten



## RData Files aus Server runtergeladen, Objecte geladen 


- `RNA_Seq_RPKM_and_profiles.RData` --> (contains: `rna.seq.exp.profils` + `rpkm.rna.seq.counts`)
- `codingSequences.RData` --> (aet.cds, aly.cds, ath.cds, bra.cds, chi.cds, cru.cds, esa.cds, tpa.cds, all.cds)

<br>

- `ExpressionProfileDistances.RData` --> (contains: )
- `orthologsTandems.RData` --> 
- `GeneGroups.RData` --> 
- `families.RData` --> 

<br>

- BUSTED_Results
- cafe_result

- ExpressionProfileDistanceDistributions
- fubar_results

- rpkmExpressionProfiles
- tandemsDf
- fam.pos.sel
- families.exp


## `1. Load tissue specific expression data`

### 1.1 Description Section-1

`RPKM_counts_table.tsv` (id / tissue / expression) (Input)



### 1.2 Input/Output

- exec: loadRpkmRnaSeqCounts.R, loadCodingSequences.R

---
<br>

exec/loadRpkmRnaSeqCounts.R <-- `RPKM_counts_table.tsv` (id / tissue / expression) (Input)

`rna.seq.exp.profils`

exec/loadRpkmRnaSeqCounts.R --> `RNA_Seq_RPKM_and_profiles.RData` (Output) (contains: `rna.seq.exp.profils` + `rpkm.rna.seq.counts`)
                            --> save rna.seq.exp.profils as `RNA_Seq_RPKM_and_profiles.tsv`

- description to loadRpkmRnaSeqCounts.R --> roxygen2 file

<br>

---
<br>

exec/loadCodingSequences.R <-- Input: `aet.fa aly.fa ath.fa bra.fa chi.fa cru.fa esa.fa tpa.fa`  (DNA-Fasta-Files, for 8 Brassicaceaen genomes)

`all.cds`-> list containing (aet.cds, aly.cds, ath.cds, bra.cds, chi.cds, cru.cds, esa.cds, tpa.cds)

exec/loadCodingSequences.R --> `codingSequences.RData` (Output) (aet.cds, aly.cds, ath.cds, bra.cds, chi.cds, cru.cds, esa.cds, tpa.cds, all.cds)

- description to loadCodingSequences.R --> roxygen2 file



### 1.3 Datentypen, Datenstruktur

- Ãœbersicht Files for Section-1:

- `RPKM_counts_table.tsv` --> `rpkm.rna.seq.counts`
- `RNA_Seq_RPKM_and_profiles.tsv` --> `rna.seq.exp.profils`
- `RNA_Seq_RPKM_and_profiles.RData` --> (contains: `rna.seq.exp.profils` + `rpkm.rna.seq.counts`)

- `aet.fa aly.fa ath.fa bra.fa chi.fa cru.fa esa.fa tpa.fa` --> (DNA-Fasta-Files, for 8 Brassicaceaen genomes)
- `all.cds` 
- `codingSequences.RData` --> (contains: `aet.cds, aly.cds, ath.cds, bra.cds, chi.cds, cru.cds, esa.cds, tpa.cds, all.cds`)

<br>

---
<br>

- `RPKM_counts_table.tsv`  --> `rpkm.rna.seq.counts`

<br>

```md
> head(`RPKM_counts_table.tsv`)

# id                 tissue               expression

# 1 AT1G01010        cotyledon            0.20638414
# 2 AT1G01010        developing leaf      0.08695488
# 3 AT1G01010        seedling             0.27762869
# 4 AT1G01010        flower stage 16      0.07009093
# 5 AT1G01010        flower stage 9       0.35894135
# 6 AT1G01020        cotyledon            0.16516131

> str(`RPKM_counts_table.tsv`) 

# 'data.frame':   284370 obs. of  3 variables:

# $ id        : chr  "AT1G01010" "AT1G01010" "AT1G01010" "AT1G01010" ...
# $ tissue    : chr  "cotyledon" "developing leaf" "seedling" "flower stage 16" ...
# $ expression: num  0.2064 0.087 0.2776 0.0701 0.3589 ...

```
- discription to `RPKM_counts_table.tsv`
- expression ->
- tissue -> 
- id ->

---
<br>

- `RNA_Seq_RPKM_and_profiles.tsv` --> `rna.seq.exp.profils`

<br>

```md
> head(`RNA_Seq_RPKM_and_profiles.tsv`)

#     cotyledon      developing.leaf  seedling       flower.stage.16  flower.stage.9  gene

# 1 0.206384145      0.08695488       0.2776287      0.07009093       0.3589414       AT1G01010
# 2 0.165161313      0.10448020       0.2267613      0.08841315       0.4151840       AT1G01020
# 3 0.086191341      0.15355937       0.1500006      0.07077711       0.5394716       AT1G01030
# 4 0.330792887      0.09427257       0.1829741      0.08449369       0.3074667       AT1G01040
# 5 0.079472192      0.11188537       0.1738841      0.34547414       0.2892842       AT1G01050
# 6 0.003454131      0.08389586       0.5113684      0.27232766       0.1289540       AT1G01060

> str(`RNA_Seq_RPKM_and_profiles.tsv`)

# 'data.frame':   56874 obs. of  6 variables:

# $ cotyledon      : num  0.2064 0.1652 0.0862 0.3308 0.0795 ...
# $ developing.leaf: num  0.087 0.1045 0.1536 0.0943 0.1119 ...
# $ seedling       : num  0.278 0.227 0.15 0.183 0.174 ...
# $ flower.stage.16: num  0.0701 0.0884 0.0708 0.0845 0.3455 ...
# $ flower.stage.9 : num  0.359 0.415 0.539 0.307 0.289 ...
# $ gene           : chr  "AT1G01010" "AT1G01020" "AT1G01030" "AT1G01040" ...
 
```
<br>

- `RNA_Seq_RPKM_and_profiles.RData` --> (contains: `rna.seq.exp.profils` + `rpkm.rna.seq.counts`)

- discription to `RNA_Seq_RPKM_and_profiles.tsv`
- 

---
<br>

- `aet.fa aly.fa ath.fa bra.fa chi.fa cru.fa esa.fa tpa.fa` --> (DNA-Fasta-Files, for 8 Brassicaceaen genomes)

- `all.cds` 

```md




```
<br>

- `codingSequences.RData` --> (contains: `aet.cds, aly.cds, ath.cds, bra.cds, chi.cds, cru.cds, esa.cds, tpa.cds, all.cds`)

<br>





## `2. Compute pairwise gene expression profile distances within gene groups`

### 2.1 Description Section-2

<br>

### 2.2 Input/Output

- exec: 
- `computeExpressionProfileDistances.R` 
- `LoadOrthologsAndTandems.R`
- `defineGeneSets.R`
- `loadGeneFamilies.R`

<br>

- R-functions:
- `expression_funks.R`

<br>

---
<br>

exec/`computeExpressionProfileDistances.R` <-- `tandems.lst`, `orthologs.lst`, `families.lst`, `expressionProfilesDists` (Input)

<br>

- Objects/Variables generated: 

- `families.exp.prof.dists` & `families.exp.prof.dists.tissue`
- `orthologs.exp.prof.dists` & `orthologs.exp.prof.dists.tissue`
- `tandems.exp.prof.dists` & `tandems.exp.prof.dists.tissue`

<br>

exec/`computeExpressionProfileDistances.R` --> `ExpressionProfileDistances.RData` (Output) (contains the above Objects/variables)

<br>

---
<br>

- fehlende Inputs: `tandems.lst` (list), `orthologs.lst` (list), `families.lst` (list), `expressionProfilesDists`

<br>

---
<br>

exec/`LoadOrthologsAndTandems.R` <-- `eight_brassicaceae_tandems.txt`, `eight_brassicaceae_orthologs.txt`

- Objects/Variables generated: `orthologs` & `tandems`

exec/`LoadOrthologsAndTandems.R` --> `orthologsTandems.RData` (contains the above Objects/variables)

<br>

---
<br>

exec/`defineGeneSets.R` <-- `orthologs` & `tandems`

- Objects/Variables generated: `orthologs.lst` & `tandems.lst`

exec/`defineGeneSets.R` --> `GeneGroups.RData` (contains the above Objects/variables)

<br>

---
<br>

exec/`loadGeneFamilies.R` <-- `mcl_output.txt` , `mcl_table.tsv` (from markov clustering tool mcl)

- Objects/Variables generated: `families.lst` & (families.genes.df, families.df) (nicht essentiel)

exec/`loadGeneFamilies.R` --> `families.RData` (contains the above Objects/variables)

<br>

---
<br>

R-functions/`expression_funks.R` <-- gene.accessions (??), rna.seq.exp.profils (aus Section-1)

- Objects/Variables generated: `expressionProfilesDists`

R-functions/`expression_funks.R` --> `expressionProfilesDists`

<br>

---

### 2.3 Datentypen, Datenstruktur

<br>

- `tandems.lst`, `orthologs.lst`, `families.lst`, `expressionProfilesDists` (Inputs)
- `families.exp.prof.dists` & `families.exp.prof.dists.tissue`
- `orthologs.exp.prof.dists` & `orthologs.exp.prof.dists.tissue`
- `tandems.exp.prof.dists` & `tandems.exp.prof.dists.tissue`
- `eight_brassicaceae_tandems.txt`, `eight_brassicaceae_orthologs.txt`
- `orthologs` & `tandems`
- `mcl_output.txt` , `mcl_table.tsv`

<br>

---
<br>

- `tandems.lst` , `orthologs.lst` , `families.lst`

<br>

```md

> str(`tandems.lst`)

List of 3266 (Key: Value pairs)

# $ tandem_cluster_1   : chr [1:152] "CARHR230400.1" "CARHR230340.1" "Carubv10004324m" "Bra013677" ...
# $ tandem_cluster_2   : chr [1:125] "Bra007390" "Bra007392" "Bra014571" "Thhalv10005920m" ...
# $ tandem_cluster_3   : chr [1:137] "Thhalv10001065m" "Thhalv10001146m" "Thhalv10001152m" "Tp2g10040" ...
# $ tandem_cluster_4   : chr [1:96] "475793" "CARHR059360.1" "CARHR059200.1" "908232" ...
# $ tandem_cluster_5   : chr [1:126] "Thhalv10015283m" "Thhalv10015331m" "Thhalv10015353m" "Thhalv10015652m" ...
# $ tandem_cluster_6   : chr [1:101] "aet_11297" "aet_11303" "CARHR050860.1" "475168" ...
...

> str(`orthologs.lst`)

List of 10111 (Key: Value pairs)

# $ ortholog_cluster_1    : Named chr [1:8] "311263" "AT1G01790" "Carubv10011770m" "CARHR000530" ...
  ..- attr(*, "names")= chr [1:8] "aly" "ath" "cru" "chi" ...

# $ ortholog_cluster_2    : Named chr [1:8] "311286" "AT1G01510" "Carubv10011468m" "CARHR000860" ...
  ..- attr(*, "names")= chr [1:8] "aly" "ath" "cru" "chi" ...

# $ ortholog_cluster_3    : Named chr [1:8] "311290" "AT1G01460" "Carubv10009221m" "CARHR000910" ...
  ..- attr(*, "names")= chr [1:8] "aly" "ath" "cru" "chi" ...

# $ ortholog_cluster_4    : Named chr [1:8] "311315" "AT1G01190" "Carubv10008749m" "CARHR001280" ...
  ..- attr(*, "names")= chr [1:8] "aly" "ath" "cru" "chi" ...

# $ ortholog_cluster_5    : Named chr [1:8] "311338" "AT1G02280" "Carubv10010930m" "CARHR001580" ...
  ..- attr(*, "names")= chr [1:8] "aly" "ath" "cru" "chi" ...
...

> length(`families.lst`)   
[1] 23716

```
- description

---
<br>

- `families.exp.prof.dists` & `families.exp.prof.dists.tissue`
- `orthologs.exp.prof.dists` & `orthologs.exp.prof.dists.tissue`
- `tandems.exp.prof.dists` & `tandems.exp.prof.dists.tissue`

<br>

```md
> str(`families.exp.prof.dists`)

List of 16910 (Key: Value pairs)

# $ cluster_1    : 'dist' num [1:12880] 0.61 0.962 0.436 0.681 0.336 ...

  ..- attr(*, "Size")= int 161
  ..- attr(*, "Labels")= chr [1:161] "AT1G06180" "AT1G08810" "AT1G09540" "AT1G16490" ...
  ..- attr(*, "Diag")= logi FALSE
  ..- attr(*, "Upper")= logi FALSE
  ..- attr(*, "method")= chr "euclidean"
  ..- attr(*, "call")= language dist(x = exp.profs[, tissues], method = dist.method)

# $ cluster_2    : 'dist' num [1:16110] 0.936 NA NA NA NA ...

  ..- attr(*, "Size")= int 180
  ..- attr(*, "Labels")= chr [1:180] "AT1G10890" "AT1G11810" "AT1G24800" "AT1G24881" ...
  ..- attr(*, "Diag")= logi FALSE
  ..- attr(*, "Upper")= logi FALSE
  ..- attr(*, "method")= chr "euclidean"
  ..- attr(*, "call")= language dist(x = exp.profs[, tissues], method = dist.method)

# $ cluster_3    : 'dist' num [1:8646] 0.396 0.342 0.648 0.706 0.635 ...

  ..- attr(*, "Size")= int 132
  ..- attr(*, "Labels")= chr [1:132] "AT1G11340" "AT1G11410" "AT1G61610" "AT3G45860" ...
  ..- attr(*, "Diag")= logi FALSE
  ..- attr(*, "Upper")= logi FALSE
  ..- attr(*, "method")= chr "euclidean"
  ..- attr(*, "call")= language dist(x = exp.profs[, tissues], method = dist.method)

# $ cluster_4    : 'dist' num [1:4278] 0.993 0.906 0.992 0.993 0.973 ...

  ..- attr(*, "Size")= int 93
  ..- attr(*, "Labels")= chr [1:93] "AT1G11070" "AT1G12040" "AT1G21310" "AT1G23720" ...
  ..- attr(*, "Diag")= logi FALSE
  ..- attr(*, "Upper")= logi FALSE
  ..- attr(*, "method")= chr "euclidean"
  ..- attr(*, "call")= language dist(x = exp.profs[, tissues], method = dist.method)
...

> str(`families.exp.prof.dists.tissue`)

List of 16910 

 $ cluster_1    :List of 5

  ..$ cotyledon      : 'dist' num [1:12880] 0.1874 0.0163 0.0128 0.5367 0.1525 ...
  .. ..- attr(*, "Size")= int 161
  .. ..- attr(*, "Labels")= chr [1:161] "AT1G06180" "AT1G08810" "AT1G09540" "AT1G16490" ...
  .. ..- attr(*, "Diag")= logi FALSE
  .. ..- attr(*, "Upper")= logi FALSE
  .. ..- attr(*, "method")= chr "euclidean"
  .. ..- attr(*, "call")= language dist(x = setNames(exp.profs[, tissue], rownames(exp.profs)), method = dist.method)

  ..$ developing leaf: 'dist' num [1:12880] 0.304 0.342 0.337 0.326 0.147 ...
  .. ..- attr(*, "Size")= int 161
  .. ..- attr(*, "Labels")= chr [1:161] "AT1G06180" "AT1G08810" "AT1G09540" "AT1G16490" ...
  .. ..- attr(*, "Diag")= logi FALSE
  .. ..- attr(*, "Upper")= logi FALSE
  .. ..- attr(*, "method")= chr "euclidean"
  .. ..- attr(*, "call")= language dist(x = setNames(exp.profs[, tissue], rownames(exp.profs)), method = dist.method)

  ..$ seedling       : 'dist' num [1:12880] 0.0389 0.1536 0.0444 0.078 0.1069 ...
  .. ..- attr(*, "Size")= int 161
  .. ..- attr(*, "Labels")= chr [1:161] "AT1G06180" "AT1G08810" "AT1G09540" "AT1G16490" ...
  .. ..- attr(*, "Diag")= logi FALSE
  .. ..- attr(*, "Upper")= logi FALSE
  .. ..- attr(*, "method")= chr "euclidean"
  .. ..- attr(*, "call")= language dist(x = setNames(exp.profs[, tissue], rownames(exp.profs)), method = dist.method)

  ..$ flower stage 16: 'dist' num [1:12880] 0.3857 0.8275 0.2403 0.0392 0.2113 ...
  .. ..- attr(*, "Size")= int 161
  .. ..- attr(*, "Labels")= chr [1:161] "AT1G06180" "AT1G08810" "AT1G09540" "AT1G16490" ...
  .. ..- attr(*, "Diag")= logi FALSE
  .. ..- attr(*, "Upper")= logi FALSE
  .. ..- attr(*, "method")= chr "euclidean"
  .. ..- attr(*, "call")= language dist(x = setNames(exp.profs[, tissue], rownames(exp.profs)), method = dist.method)

  ..$ flower stage 9 : 'dist' num [1:12880] 0.308 0.316 0.128 0.25 0.11 ...
  .. ..- attr(*, "Size")= int 161
  .. ..- attr(*, "Labels")= chr [1:161] "AT1G06180" "AT1G08810" "AT1G09540" "AT1G16490" ...
  .. ..- attr(*, "Diag")= logi FALSE
  .. ..- attr(*, "Upper")= logi FALSE
  .. ..- attr(*, "method")= chr "euclidean"
  .. ..- attr(*, "call")= language dist(x = setNames(exp.profs[, tissue], rownames(exp.profs)), method = dist.method)

  $ cluster_2    :List of 5
  ..$ cotyledon      : 'dist' num [1:16110] 0.179 NA NA NA NA ...
  .. ..- attr(*, "Size")= int 180
  .. ..- attr(*, "Labels")= chr [1:180] "AT1G10890" "AT1G11810" "AT1G24800" "AT1G24881" ...
  .. ..- attr(*, "Diag")= logi FALSE
  .. ..- attr(*, "Upper")= logi FALSE
  .. ..- attr(*, "method")= chr "euclidean"
  .. ..- attr(*, "call")= language dist(x = setNames(exp.profs[, tissue], rownames(exp.profs)), method = dist.method)

  ..$ developing leaf: 'dist' num [1:16110] 0.0408 NA NA NA NA ...
  .. ..- attr(*, "Size")= int 180
  .. ..- attr(*, "Labels")= chr [1:180] "AT1G10890" "AT1G11810" "AT1G24800" "AT1G24881" ...
  .. ..- attr(*, "Diag")= logi FALSE
  .. ..- attr(*, "Upper")= logi FALSE
  .. ..- attr(*, "method")= chr "euclidean"
  .. ..- attr(*, "call")= language dist(x = setNames(exp.profs[, tissue], rownames(exp.profs)), method = dist.method)

  ..$ seedling       : 'dist' num [1:16110] 0.158 NA NA NA NA ...
  .. ..- attr(*, "Size")= int 180
  .. ..- attr(*, "Labels")= chr [1:180] "AT1G10890" "AT1G11810" "AT1G24800" "AT1G24881" ...
  .. ..- attr(*, "Diag")= logi FALSE
  .. ..- attr(*, "Upper")= logi FALSE
  .. ..- attr(*, "method")= chr "euclidean"
  .. ..- attr(*, "call")= language dist(x = setNames(exp.profs[, tissue], rownames(exp.profs)), method = dist.method)

  ..$ flower stage 16: 'dist' num [1:16110] 0.8 NA NA NA NA ...
  .. ..- attr(*, "Size")= int 180
  .. ..- attr(*, "Labels")= chr [1:180] "AT1G10890" "AT1G11810" "AT1G24800" "AT1G24881" ...
  .. ..- attr(*, "Diag")= logi FALSE
  .. ..- attr(*, "Upper")= logi FALSE
  .. ..- attr(*, "method")= chr "euclidean"
  .. ..- attr(*, "call")= language dist(x = setNames(exp.profs[, tissue], rownames(exp.profs)), method = dist.method)

  ..$ flower stage 9 : 'dist' num [1:16110] 0.422 NA NA NA NA ...
  .. ..- attr(*, "Size")= int 180
  .. ..- attr(*, "Labels")= chr [1:180] "AT1G10890" "AT1G11810" "AT1G24800" "AT1G24881" ...
  .. ..- attr(*, "Diag")= logi FALSE
  .. ..- attr(*, "Upper")= logi FALSE
  .. ..- attr(*, "method")= chr "euclidean"
  .. ..- attr(*, "call")= language dist(x = setNames(exp.profs[, tissue], rownames(exp.profs)), method = dist.method)

  ...

```





## `3. Expression Profile based Function Diversity`

### 3.1 Description

### 3.2 Input/Output

### 3.3 Datentypen, Datenstruktur

## `4. Expression vector space analysis`






Kommentare: Beschreiben Sie Funktionen und wichtige Codeabschnitte.

Roxygen2: Verwenden Sie Roxygen2 fÃ¼r eine formale Funktionsdokumentation.



- eventuell roxygen2 beschreibungen importiere